// material-ui-player/dist/components/VolumeBar.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VolumeBar = void 0;
var react_1 = __importStar(require("react"));
var IconButton_1 = __importDefault(require("@mui/material/IconButton"));
var Slider_1 = __importDefault(require("@mui/material/Slider"));
var styles_1 = require("@mui/material/styles");
var createStyles_1 = __importDefault(require("@mui/styles/createStyles"));
var makeStyles_1 = __importDefault(require("@mui/styles/makeStyles"));
var utils_1 = require("../lib/utils");
var icons_1 = require("../icons");
var useStyles = (0, makeStyles_1.default)(function (theme) {
    return (0, createStyles_1.default)({
        root: {
            display: 'flex',
            flexDirection: 'row',
            alignItems: 'center',
            maxWidth: theme.spacing(40),
            marginRight: theme.spacing(2)
        },
    });
});
var VolumeSlider = (0, styles_1.styled)(function (props) { return react_1.default.createElement(Slider_1.default, __assign({}, props)); })(function (_a) {
    var theme = _a.theme, thickness = _a.thickness, color = _a.color, thumb = _a.thumb;
    return (__assign({}, (0, utils_1.getSliderSizes)(thickness, ({
        thumb: {
            color: theme.palette[color || 'primary'].main,
            display: thumb === false ? 'none' : 'block',
        },
        rail: {
            color: theme.palette[color || 'primary'].main,
        },
        track: {
            color: theme.palette[color || 'primary'].main,
        },
    }))));
});
var VolumeBar = function (props) {
    var _a, _b;
    var mute = props.mute, player = props.player, MuteProps = props.MuteProps, styles = __rest(props, ["mute", "player", "MuteProps"]);
    var _c = (0, react_1.useState)(100), volumeBar = _c[0], setVolumeBar = _c[1];
    var _d = (0, react_1.useState)(false), muted = _d[0], setMuted = _d[1];
    var classes = useStyles();
    var onVolumeChange = function (_, newValue) {
        var volume = newValue;
        setVolumeBar(volume);
        player.volume = volume / 100;
    };
    var onMuteClick = function () {
        var newState = !muted;
        setMuted(newState);
        player.muted = newState;
    };
    return (react_1.default.createElement("div", { className: classes.root },
        mute &&
            react_1.default.createElement(IconButton_1.default, __assign({ color: styles.color || 'primary', "aria-label": "Mute", onClick: onMuteClick }, MuteProps === null || MuteProps === void 0 ? void 0 : MuteProps.attributes, { size: "large" }), muted ?
                ((_a = MuteProps === null || MuteProps === void 0 ? void 0 : MuteProps.icons) === null || _a === void 0 ? void 0 : _a[0]) || react_1.default.createElement(icons_1.VolumeOff, null) :
                ((_b = MuteProps === null || MuteProps === void 0 ? void 0 : MuteProps.icons) === null || _b === void 0 ? void 0 : _b[1]) || react_1.default.createElement(icons_1.VolumeUp, null)),
        react_1.default.createElement(VolumeSlider, __assign({ value: volumeBar, onChange: onVolumeChange }, styles))));
};
exports.VolumeBar = VolumeBar;
